internal/core/processor/evolution/client.go

```diff
@@ -0,0 +1,141 @@
+	"fmt"
+	"log/slog"
+	"math/big"
+	"net/http"
+)
+
+const (
+	jsonContentType = "application/json"
+	postMethod      = "POST"
+)
+
+// Define an interface for HTTP client operations
+type LaosRPCRequests interface {
+	LatestFinalizedBlockHash() (string, error)
+	BlockNumber(blockHash string) (*big.Int, error)
+}
+
+// Define an interface for HTTP client operations
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type LaosHTTP struct {
+	Client HTTPClient
+	Url    string
+}
+
+func NewLaosHTTP(url string) LaosRPCRequests {
+	return LaosHTTP{
+		Client: &http.Client{},
+		Url:    url,
+	}
+}
+
+// BlockHeader struct to represent the block header in the JSON response
+type BlockHeader struct {
+	ParentHash     string `json:"parentHash"`
+	Number         string `json:"number"`
+	StateRoot      string `json:"stateRoot"`
+	ExtrinsicsRoot string `json:"extrinsicsRoot"`
+}
+
+// Block struct to represent the block in the JSON response
+type Block struct {
+	Header BlockHeader `json:"header"`
+}
+
+// Response struct to represent the entire JSON response
+type ChainGetBlock struct {
+	JSONRPC string `json:"jsonrpc"`
+	Result  struct {
+		Block Block `json:"block"`
+	} `json:"result"`
+	ID int `json:"id"`
+}
+
+// Response struct to represent the entire JSON response
+type ChainGetFinalizedHead struct {
+	JSONRPC string `json:"jsonrpc"`
+	Result  string `json:"result"`
+	ID      int    `json:"id"`
+}
+
+func (l LaosHTTP) makeRequest(payload []byte, result interface{}) error {
+	req, err := http.NewRequest(postMethod, l.Url, bytes.NewBuffer(payload))
+	if err != nil {
+		return fmt.Errorf("error creating request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", jsonContentType)
+
+	resp, err := l.Client.Do(req)
+	if err != nil {
+		return fmt.Errorf("error making request to %s: %w", l.Url, err)
+	}
+
+	defer func() {
+		err := resp.Body.Close()
+		if err != nil {
+			slog.Error("error closing response body", "err", err)
+		}
+	}()
+
+	if resp.StatusCode >= http.StatusBadRequest {
+		return fmt.Errorf("error in request to %s, got status code: %d", l.Url, resp.StatusCode)
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
+		return fmt.Errorf("error decoding JSON: %w", err)
+	}
+
+	return nil
+}
+
+func (l LaosHTTP) LatestFinalizedBlockHash() (string, error) {
+	payload := []byte(`{"jsonrpc":"2.0","method":"chain_getFinalizedHead","params":[],"id":1}`)
+
+	var response ChainGetFinalizedHead
+	if err := l.makeRequest(payload, &response); err != nil {
+		return "", err
+	}
+
+	return response.Result, nil
+}
+
+func (l LaosHTTP) BlockNumber(blockHash string) (*big.Int, error) {
+	payload := []byte(fmt.Sprintf(`{"jsonrpc":"2.0","method":"chain_getBlock","params":[%q],"id":1}`, blockHash))
+
+	var response ChainGetBlock
+	if err := l.makeRequest(payload, &response); err != nil {
+		return nil, err
+	}
+
+	hexNumber := response.Result.Block.Header.Number
+
+	blockNumber, err := hexToDecimal(hexNumber)
```

Comment:
you are wrapping this error two times, here and in the function. Is this necessary?